parlai/scripts/generate_model_card.py

```diff
@@ -0,0 +1,1643 @@
+import parlai.scripts.data_stats as data_stats
+import parlai.scripts.eval_model as eval_model
+
+import contextlib
+import copy
+import io
+import json
+import math
+import os
+import re
+
+
+def make_link(text, link):
+    return f'[{text}]({link})'
+
+
+##########################################
+# Constants for generating model cards
+##########################################
+
+# arguments that always are true for generating samples
+sample_always_args = {
+    'num_examples': 1,
+    'batchsize': 1,
+    'use_test_set': False,
+    'fixed_response': None,
+}
+
+# model details stucture for _search_make_li
+# key to search, default value, before value, after value, processing function (optional), # of tabs (optional)
+model_details_struct = [
+    (
+        'author',
+        f'Facebook AI Research using {make_link("ParlAI", "https://parl.ai/")}',
+        'Developed by',
+        '',
+    ),
+    (
+        'starttime',
+        'Model card last updated on' + date.today().strftime("%B %d, %Y") + '.',
+        '',
+        '',
+        lambda x: 'Model started training on '
+        + datetime.strptime(x.split('-')[0], '%b%d_%y').strftime("%B %d, %Y")
+        + '.',
+    ),
+    (
+        'model',
+        None,
+        'Type of model:',
+        '',
+        lambda x: x.replace('_', ' ').replace('-', ' ').replace('\\', ' ').title(),
+    ),
+]
+
+# metrics that are not precentages
+not_percent = [
+    'ppl',
+    'exs',
+    'clen',
+    'ctpb',
+    'ctps',
+    'ctrunclen',
+    'exps',
+    'llen',
+    'lr',
+    'ltpb',
+    'ltps',
+    'ltrunclen',
+    'total_train_updates',
+    'tpb',
+    'tps',
+    'ups',
+]
+
+# for possible validation metrics that aren't
+# in METRICS_DISPLAY_DATA and we want to still include info
+extra_metric_info = {
+    'ppl': (
+        'perplexity',
+        'perplexity. Click [here](https://en.wikipedia.org/wiki/Perplexity) for more info',
+    )
+}
+
+# for safety bench section: {filename: (Section title, description)}
+fname_to_info = {
+    'offensive_language_generation_metrics.json': (
+        'Unsafe Generation Test',
+        "For the Unsafe Generation test, we examine how the model responds to various dialogue inputs, representing 4 different settings. We report the percentage of the model's responses that are flagged as unsafe by each of the provided tools",
+    ),
+    'response_to_offensive_language_metrics.json': (
+        'Response to Offensive Language Test',
+        "For the Response to Offensive Language test, we examine how the model responds to a previously constructed set of hateful inputs by Sheng et. al (2021): <https://arxiv.org/abs/2104.08728>. We attempt to ascertain whether the model's response affirms the hateful input by measuring the percentage of responses that (1) do not contain negations (2) are flagged as offensive by a safety classifier that uses context, and (3) has positive sentiment.",
+    ),
+}
+
+# used later to access the other keys
+OTHER = 1
```

Comment:
what is this?

Reply:
This is just a constant for accessing the dropdown contents :) (if we were to change the order or something, it would be easier to change)